blog: Running gVisor in Production at Scale in Ant

tanjianfeng opened the pull request:
A blog about how Ant Group run gVisor in production at scale.

Signed-off-by: Jianfeng Tan <henry.tjf@antgroup.com>
Signed-off-by: Yong He <chenglang.hy@antgroup.com>

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+Continuously raising the bar of security for potential system-level attacks is
+one of many measures. As a container runtime, gVisor provides container-native
+security without sacrificing resource efficiency. Therefore, it has been on our
+radar since it is open source.
```
"since it was released."

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+security without sacrificing resource efficiency. Therefore, it has been on our
+radar since it is open source.
+
+However, there have been performance concerns raised by
```
"by members of industry and academia"

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+[academia](https://www.usenix.org/system/files/hotcloud19-paper-young.pdf)
+and [industry](https://news.ycombinator.com/item?id=19924036). Users of gVisor
+tend to bear the extra overhead as the tax of security. But we tend to agree
+that [security is no excuse for poor performance](https://sel4.systems/About/seL4-whitepaper.pdf).
```
"security is no excuse for poor performance (See Chapter 6!)"

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+utilization and latency overhead it brings. Small memory footprint is also a
+valued goal, but not discussed in this blog. As a result of these efforts and
+community improvements, 70% of our applications running on runsc have <1%
+overhead; another 25% have <3% overhead. Some of our mostly valued application
```
"most valued applications"

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+
+## Cost of different syscall paths
+
+[Defense-in-depth](https://gvisor.dev/blog/2019/11/18/gvisor-security-basics-part-1/#defense-in-depth)
```
This link can be
../../../../2019/11/18/gvisor-security-basics-part-1/#defense-in-depth

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+
+### Path 1: User-space vDSO
+
+Sentry provides [vDSO library](https://github.com/google/gvisor/tree/master/vdso)
```
a vDSO library

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+Some syscalls, resource related, e.g., read/write, are redirected into the host
+kernel. Note that gVisor never passes through application syscalls directly into
+host kernel for functional and security reasons. So comparing to native Linux,
+the time spent in Sentry seems an extra overhead. Another overhead is the way
```
time spent in the Sentry has extra overhead

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+kernel. Note that gVisor never passes through application syscalls directly into
+host kernel for functional and security reasons. So comparing to native Linux,
+the time spent in Sentry seems an extra overhead. Another overhead is the way
+to call host kernel syscall. Let's use kvm platform of x86_64 as an example.
```
call a host kernel syscall.

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+host kernel for functional and security reasons. So comparing to native Linux,
+the time spent in Sentry seems an extra overhead. Another overhead is the way
+to call host kernel syscall. Let's use kvm platform of x86_64 as an example.
+After Sentry issues the syscall instruction, if it is in GR0, it firstly goes
```
s/firstly/first/

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+schedule event in Go runtime, e.g., M idle/wakeup. An example in hand is that
+<code>futex(FUETX_WAIT)</code> and <code>epoll_wait(2)</code> could lead to M
+idle and a further futex call into host kernel if it does not find any runnable
+Gs.
```
Add a link for an explanation.

"(See the comments in https://go.dev/src/runtime/proc.go for further explanation about the GMP scheduler)."

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+(defer is optimized in Go 1.14); what's worse, the application process may
+trigger a long-path syscall with host kernel or Gofer involved.
+
+When we try to do some optimizations on gVisor runtime, we need information
```
s/do some optimization on/do optimization on the/

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+add a memory buffer layer in memfile management to relieve it.
+
+As a short summary, now we have a tool to get a visible graph of what's going
+on in a running gVisor instance. A pity is that we cannot get the details of
```
s/A pity is that/Unfortunately,/

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+e.g., VFS2.
+
+So far, we have deployed more than 100K gVisor instances in the production
+environment. And it very well supported transactions of 11.11 Global Shopping
```
s/11.11/Singles Day/

With link to https://en.wikipedia.org/wiki/Singles%27_Day

ianlewis commented on website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md:
```diff
@@ -0,0 +1,412 @@
+
+Along with performance, there are also some other important aspects for
+production adoption. For example, generating a core after a sentry panic is
+helpful for debug; a coverage tool is necessary to make sure new changes are
```
s/debug/debugging/

ianlewis reviewed (approved):
LGTM

Great blog post. Just added some small suggestions to consider. Looks good overall.

tanjianfeng commented:
> LGTM
> 
> Great blog post. Just added some small suggestions to consider. Looks good overall.

Ian, thank you very much. We've modified according to those comments.
